pkg/compression/pool.go

```diff
@@ -0,0 +1,354 @@
+package compression
+
+import (
+	"bufio"
+	"io"
+	"runtime"
+	"sync"
+
+	"github.com/golang/snappy"
+	"github.com/klauspost/compress/flate"
+	"github.com/klauspost/compress/gzip"
+	"github.com/klauspost/compress/zstd"
+	"github.com/pierrec/lz4/v4"
+)
+
+// WriterPool is a pool of io.Writer
+// This is used by every chunk to avoid unnecessary allocations.
+type WriterPool interface {
+	GetWriter(io.Writer) io.WriteCloser
+	PutWriter(io.WriteCloser)
+}
+
+// ReaderPool similar to WriterPool but for reading chunks.
+type ReaderPool interface {
+	GetReader(io.Reader) (io.Reader, error)
+	PutReader(io.Reader)
+}
+
+var (
+	// Gzip is the gnu zip compression pool
+	Gzip     = GzipPool{level: gzip.DefaultCompression}
+	Lz4_64k  = LZ4Pool{bufferSize: 1 << 16} // Lz4_64k is the l4z compression pool, with 64k buffer size
+	Lz4_256k = LZ4Pool{bufferSize: 1 << 18} // Lz4_256k uses 256k buffer
+	Lz4_1M   = LZ4Pool{bufferSize: 1 << 20} // Lz4_1M uses 1M buffer
+	Lz4_4M   = LZ4Pool{bufferSize: 1 << 22} // Lz4_4M uses 4M buffer
+	Flate    = FlatePool{}
+	Zstd     = ZstdPool{}
+	// Snappy is the snappy compression pool
+	Snappy SnappyPool
+	// Noop is the no compression pool
+	Noop NoopPool
+)
+
+func GetWriterPool(enc Encoding) WriterPool {
+	return GetReaderPool(enc).(WriterPool)
+}
+
+func GetReaderPool(enc Encoding) ReaderPool {
+	switch enc {
+	case EncGZIP:
+		return &Gzip
+	case EncLZ4_64k:
+		return &Lz4_64k
+	case EncLZ4_256k:
+		return &Lz4_256k
+	case EncLZ4_1M:
+		return &Lz4_1M
+	case EncLZ4_4M:
+		return &Lz4_4M
+	case EncSnappy:
+		return &Snappy
+	case EncNone:
+		return &Noop
+	case EncFlate:
+		return &Flate
+	case EncZstd:
+		return &Zstd
+	default:
+		panic("unknown encoding")
+	}
```

Comment:
nit: None* of the pools are referenced directly outside of this package, so it's safe to unexport all of them and just leave `GetReaderPool`/`GetWriterPool` as the public API for pools. 

Personally I'm normally a fan of unexporting as much as possible as it reduces the surface where good API documentation is necessary. I don't feel too strongly about this at the moment, so I don't mind if this doesn't get changed. 

*: `compression.Gzip` is referenced a few times, but all references can be replaced with `compression.GetReaderPool(compression.EncGZIP)` or `compression.GetWriterPool(compression.EncGZIP)`.